rfd/xxxx-tsh-bench-db.md

```diff
@@ -0,0 +1,166 @@
+  For example, on the SSH benchmark, the servers list used to select the server
+  where the command will be executed is fetched on the suite function, and the
+  workload function only connects to it.
+
+## Details
+The database flow areas will be split into isolated suites, allowing developers
+to test and validate them individually.
+
+The suites will use an interface (`DatabaseClient`) to interact with databases.
+Decoupling it from the suites will increase reusability and also make it quicker
+to enable more database protocols to the benchmark tests.
+
+This interface is returned after connecting to the database. Each protocol will
+define its connection flow respecting the `ConnectFunc`.
+
+```go
+// DatabaseConnectionConfig contains all information necessary to establish a
+// new database connection.
+type DatabaseConnectionConfig struct {
+	// Protocol database protocol.
+	Protocol string
+	// URI direct database connection URI.
+	URI string
+	// Username database username the connection should use.
+	Username string
+	// Database database name where the connection should point to.
+	Database string
+	// ProxyAddress Teleport database proxy address.
+	ProxyAddress string
+	// TLSConfig TLS configuration containing Teleport CA and database
+	// certificates.
+	TLSConfig *tls.Config
+}
+
+// ConnectFunc is a function that establishes a database connection and returns
+// the DatabaseClient.
+type ConnectFunc func(ctx context.Context, config *DatabaseConnectionConfig) (DatabaseClient, error)
+
+// DatabaseClient represents a database connection.
+type DatabaseClient interface {
+	// Ping runs a command on the database to ensure the connection is alive.
+	Ping(ctx context.Context) error
+	// Close closes the connection.
+	Close(ctx context.Context) error
+	// Setup creates the necessary struct to create and load records.
+	Setup(ctx context.Context) error
+	// Teardown drops anything the test has created.
+	Teardown(ctx context.Context) error
+	// CreateRecord creates a new record on the database.
+	CreateRecord(ctx context.Context) error
+	// LoadRecord loads a record from the database.
+	LoadRecord(ctx context.Context) error
+}
+```
+
+#### Connection establishment
+It covers authorization capabilities and also initial connections. Although the
+other suites will also cover this, we can have more options, like providing
+different users and having the ability to run it across different database
+protocols. It will not rely on any table or struct on the connected database, as
+it will only perform a ping to ensure the connection is healthy. For databases
+without support to ping, a simple query will be sent. For example, the suite
+will send a `SELECT 1;` query on PostgreSQL databases.
+
+This suite can also validate changes on the access checker without running large
+amounts of queries.
+
+```code
+$ tsh bench db connect --db-user=postgres --db-name=postgres postgres-dev
+```
+
+#### Import/Export data
```

Comment:
The import/export bench will require db setup

I wonder if we can additionally support query benchmark that will test how db agents is behaving when receiving many query and how many query per second can handle. 

The benefit of this command is that it doesn't' require database setup and the query can be a db ping query like `select 1;`
For instance `tsh bench db query --query="select 1;" --connections=10 --time=5m [--url=]`

Reply:
> I wonder if we can additionally support query benchmark that will test how db agents is behaving when receiving many query and how many query per second can handle.

Do you mean custom queries? Because given your example of having extensive data being created/loaded will be covered by the "Import/Export data".

In addition, to speed up the case where we're issuing multiple tests in a short period (let's say during development), having a flag to keep the tables and data might be helpful.

Overall I like the idea of having a custom query command. Besides the ability to use existing structures, it will also give users a quick way to validate unique database features. However, it might only benefit SQL-compatible databases, such as MongoDB, whose protocol is not text-based. [I suggested having a command to execute external tools](https://github.com/gravitational/teleport/pull/27856#discussion_r1230954799) that use-case would cover all the databases. What do you think?

Reply:
> Do you mean custom queries? B
Yea, but the `select 1;` should be default or maybe we can sperate it by command and introduce a  `tsh bench db ping` subcommand